### PR TITLE
Ensure Directory.Delete works fine even if user has no ListDirectory permissions

### LINKED ISSUE #56922: .Net 5 Directory.Delete needs new permissions (vs .Net 4.8)
### Description
`Directory.Delete(somePath, recursive: true)` throws an `UnauthorizedAccessException` when the user has no `ListDirectory` rights on the parent directory

Repro sample:
```csharp
var parent = Directory.CreateDirectory(@"C:\Temp\someparent");
var ac = parent.GetAccessControl();
ac.SetAccessRule(new FileSystemAccessRule(WindowsIdentity.GetCurrent().User, FileSystemRights.ListDirectory, AccessControlType.Deny));
parent.SetAccessControl(ac);

var subDir = Directory.CreateDirectory(@"C:\Temp\someparent\subdir");
Directory.Delete(subDir.FullName, recursive: true);
```

throws:
```
UnauthorizedAccessException: Access to the path 'C:\Temp\someparent\subdir' is denied.
at System.IO.FileSystem.GetFindData(String fullPath, WIN32_FIND_DATA& findData)
   at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
   at System.IO.Directory.Delete(String path, Boolean recursive)
```
This is unexpected, because:
- the same user created the directory
- the user can still create sub directories and files within the directory
- same code works without exception in .Net 4.8
- the directory can be deleted without the recursive: true option

### Configuration
.NET target: Core 5.0.8, and Framework 48 (4.0.30319.42000)
Operating system: Windows 20H2

### Regression?
Yes the same code example from above works without problems in .Net 4.8

### Other information
In our case, enabling the list permission on the parent directory is not an acceptable solution, due to the security implications.

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #56922 which got introduced by https://github.com/dotnet/corefx/pull/28124/files

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs [hunk 1]
```diff
@@ -186,7 +186,10 @@ public static void RemoveDirectory(string fullPath, bool recursive)
             }
 
             Interop.Kernel32.WIN32_FIND_DATA findData = default;
-            GetFindData(fullPath, isDirectory: true, ref findData);
+            // FindFirstFile($path) (used by GetFindData) fails with ACCESS_DENIED when user has no ListDirectory rights
+            // but FindFirstFile($path/*") (used by RemoveDirectoryRecursive) works fine in such scenario.
+            // So we ignore it here and let RemoveDirectoryRecursive throw if FindFirstFile($path/*") fails with ACCESS_DENIED.
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:191]: Do you know why `Path.Join(fullPath, "*")` is needed in `RemoveDirectoryRecursive`?

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs [hunk 2]
```diff
@@ -209,6 +212,8 @@ private static void GetFindData(string fullPath, bool isDirectory, ref Interop.K
                 // File not found doesn't make much sense coming from a directory.
                 if (isDirectory && errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
                     errorCode = Interop.Errors.ERROR_PATH_NOT_FOUND;
+                if (isDirectory && errorCode == Interop.Errors.ERROR_ACCESS_DENIED && ignoreAccessDenied)
+                    return;
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:216]:
```suggestion
                if (isDirectory) 
                {
                    if (errorCode == Interop.Errors.ERROR_ACCESS_DENIED && ignoreAccessDenied)
                        return;
                        
                    // File not found doesn't make much sense coming from a directory.
                    if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
                        errorCode = Interop.Errors.ERROR_PATH_NOT_FOUND;
                }
```
nit, remove duplicated check

File: src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+using Microsoft.DotNet.XUnitExtensions;
+using System.Security.Principal;
+using System.Security.AccessControl;
+
+namespace System.IO.Tests
+{
+    public partial class Directory_Delete_str_bool : Directory_Delete_str
+    {
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
```
Comment (Jozkee) [inline src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:15]: If this is only compiling on Windows does it really need the `PlatformSpecific` attribute? Perhaps you don't need to make `Directory_Delete_str_bool` partial if you keep the attribute.

File: src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            string parentPath = GetTestFilePath();
+            var parent = Directory.CreateDirectory(parentPath);
+            var ac = parent.GetAccessControl();
+            ac.SetAccessRule(new FileSystemAccessRule(WindowsIdentity.GetCurrent().User, FileSystemRights.ListDirectory, AccessControlType.Deny));
+            parent.SetAccessControl(ac);
+
+            var subDir = parent.CreateSubdirectory("subdir");
+            File.Create(Path.Combine(subDir.FullName, GetTestFileName())).Dispose();
+            Directory.Delete(subDir.FullName, recursive: true);
+            Assert.False(subDir.Exists);
```
Comment (Jozkee) [inline src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:27]: Should we also test that the parent directory can be deleted?

**Review by jozkee — APPROVED** [review]
LGTM considering that this is good enough to mitigate the breaking change.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs (hunk 1 above)
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:191]: My understanding is that `FindFirstFile($pathToDirectory)` returns info for given directory, while `FindFirstFile($pathToDirectory/*)` returns info for the first file in given directory (which is needed here as we want to recursively remove all files and subfolders)

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs (hunk 2 above)
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:216]: I don't like this suggestion as it increases code nesting

File: src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs (hunk 3 above)
Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:15]: I forgot to call `Delete`:

https://github.com/dotnet/runtime/blob/ec07ac8709dd5676084a79193cac9498eaa427ff/src/libraries/System.IO.FileSystem/tests/Directory/Delete.cs#L229-L232

Which is overriden in subclasses that test `DirectoryInfo` as well:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.IO.FileSystem/tests/DirectoryInfo/Delete.cs#L27-L37

 fixed that in https://github.com/dotnet/runtime/pull/56996/commits/bd7f5f0ae71213e39803cec44d0a72ee7699c397

Now having this type partial makes sense.

Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:15]: Speaking of using`PlatformSpecific` attribute in files that are conditionally compiled only on certain OSes. I prefer to have the extra attribute, but I don't have a strong opinion about it. @ViktorHofer do we have some guidelines for that?

File: src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs (hunk 4 above)
Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:27]: We have other tests that verify that. In this particular test I wanted to focus on the lack of `FileSystemRights.ListDirectory`

File: src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/libraries/System.IO.FileSystem/tests/Directory/Delete.Windows.cs:15]: I agree with @adamsitnik. For such situations we usually move the platform specific code into a separate file.

**Pull request merged**